### PR TITLE
mac m1 cross compile

### PR DESCRIPTION
**Pull request by powderluv**
Add support for M1 cross compile. Also compile and test macos x86

### CODE REVIEW

File: .github/workflows/buildAndTest.yml [hunk 1]
```diff
@@ -44,9 +47,37 @@ jobs:
       with:
         cache-suffix: ${{ matrix.os }}-${{ matrix.targetarch }}-${{ matrix.llvmtype }}-${{ matrix.llvmbuildtype }}
 
+    - name: Configure llvm-cross-compile
+      # libzstd on GH Runners are only x86_64 to remove them.
+      if: matrix.targetarch == 'arm64'
+      run: |
+        sudo rm -rf /usr/local/lib/libzstd.*.dylib
```
Comment (jpienaar) [inline .github/workflows/buildAndTest.yml:54]: Would -DLLVM_ENABLE_ZSTD =OFF below work also?

Comment (powderluv) [reply .github/workflows/buildAndTest.yml:54]: Ah yes. I was copying what we were doing in IREE. 😀 I think ideally llvm upstream should check for architecture along with presence before linking. I know conda also adds it to the path inadvertently.

**Review by makslevental — APPROVED** [review]
CONGRATS

File: .github/workflows/buildAndTest.yml [hunk 2]
```diff
@@ -25,14 +25,17 @@ jobs:
           - llvmtype: binary
             llvmbuildtype: out-of-tree
           # Disable M1 builds until https://github.com/llvm/torch-mlir/issues/1094 is fixed
-          - targetarch: AArch64
+          - targetarch: arm64
+            os: ubuntu-22.04
```
Comment (sjain-stanford) [inline .github/workflows/buildAndTest.yml:29]: readability nit: maybe flip this to match the other pair of excludes?
```
- os: ubuntu-22.04
  targetarch: arm64

- os: macos-12
  targetarch: x86_64
```

**Review by sjain-stanford — APPROVED** [review]

File: .github/workflows/buildAndTest.yml (hunk 1 above)
Comment (sjain-stanford) [reply .github/workflows/buildAndTest.yml:54]: +1 on using the flag if it works.

**Review by sjain-stanford — CHANGES_REQUESTED** [review]
I think `Run torch-mlir unit tests` step is still using x86_64 build for running tests:
```
cmake --build build --target check-torch-mlir-all
```

You may want to configure this with `_${{ matrix.targetarch }}` as well.

File: .github/workflows/buildAndTest.yml (hunk 2 above)
Comment (powderluv) [reply .github/workflows/buildAndTest.yml:29]: Will address in follow on

**Review by sjain-stanford — COMMENTED** [review]
> Yeah we are building both architectures and testing only x86

This makes sense. The only consideration is the 2G ccache limit. As we're building twice in the same job / runner, we may always see low cache hit rate? Is it better to set this up cross compile build as a separate github workflow? This will also maximize parallelization (use a separate runner of which there are no restrictions I presume)? I'm hesitant in increasing the ccache limit of 2G more due to the reasons we discussed on discord.

**Comment by powderluv** [comment]
I think we can drop the macOS x86 build all together. I think we can now reduce the matrix considerably.

**Comment by sjain-stanford** [comment]
> I think we can drop the macOS x86 build all together. I think we can now reduce the matrix considerably.

You mean remove x86 build + tests for macos and only use arm64 build + tests? That would be great indeed :)

**Pull request merged**